libraries/standard/mqtt/src/mqtt.c

```diff
@@ -61,6 +64,318 @@ static int32_t sendPacket( MQTTContext_t * pContext, size_t bytesToSend )
+                                            packetId,
+                                            MQTTPuback,
+                                            MQTT_SEND );
+
+            if( newState != MQTTPublishDone )
+            {
+                status = MQTTIllegalState;
+            }
+            break;
+        case MQTTPubRecSend:
+            /* TODO: Send PubRec. Is there a function for this? */
+            //status = _sendPacket( pContext, packetId, MQTTPubrec );
+            //if( status == MQTTSuccess )
+            newState = MQTT_UpdateStateAck( pContext,
+                                            packetId,
+                                            MQTTPubrec,
+                                            MQTT_SEND );
+
+            if( newState != MQTTPubRelPending )
+            {
+                status = MQTTIllegalState;
+            }
+            break;
+        case MQTTPubRelSend:
+            /* TODO: Send PubRel. Is there a function for this? */
+            //status = _sendPacket( pContext, packetId, MQTTPubrel );
+            //if( status == MQTTSuccess )
+            newState = MQTT_UpdateStateAck( pContext,
+                                            packetId,
+                                            MQTTPubrel,
+                                            MQTT_SEND );
+
+            if( newState != MQTTPubCompPending )
+            {
+                status = MQTTIllegalState;
+            }
+            break;
+        case MQTTPubCompSend:
+            /* TODO: Send PubComp. Is there a function for this? */
+            //status = _sendPacket( pContext, packetId, MQTTPubcomp );
+            //if( status == MQTTSuccess )
+            newState = MQTT_UpdateStateAck( pContext,
+                                            packetId,
+                                            MQTTPubcomp,
+                                            MQTT_SEND );
+
+            if( newState != MQTTPublishDone )
+            {
+                status = MQTTIllegalState;
+            }
+            break;
+        default:
+            /* Nothing to send. */
+            break;
+    }
+    if( ( status == MQTTSuccess ) && ( newState != MQTTStateNull ) )
+    {
+        *pPublishState = newState;
+    }
+    return status;
+}
+
+/**
+ * @brief Handle received MQTT packet.
+ *
+ * @param[in] pContext MQTT Connection context.
+ * @param[in] pIncomingPacket Incoming packet.
+ *
+ * @return MQTTSuccess, MQTTIllegalState, or deserialization error.
+ */
+static MQTTStatus_t _handleMessage( MQTTContext_t * const pContext,
+                                    MQTTPacketInfo_t * pIncomingPacket )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPublishState_t publishRecordState = MQTTStateNull;
+    MQTTPubAckType_t ackType;
+    MQTTPublishInfo_t publishInfo;
+
+    assert( pContext != NULL );
+    assert( pIncomingPacket != NULL );
+
+    /* Determine ack type here so next switch doesn't duplicate code. */
```

Comment:
You can still combine two switch statements like this:
     switch( pIncomingPacket->type )
    {
              case MQTT_PACKET_TYPE_PUBLISH:
              /* Process Publish */
              break;
             case MQTT_PACKET_TYPE_PUBACK:
            ackType = MQTTPuback;
            /* Note no break */
            case MQTT_PACKET_TYPE_PUBREC:
            ackType = MQTTPubrec;
           
        case MQTT_PACKET_TYPE_PUBREL:
            ackType = MQTTPubrel;
            
        case MQTT_PACKET_TYPE_PUBCOMP:
            ackType = MQTTPubcomp;
      /* common ack processing code */
      break;
      default:
      /* error */

}

Reply:
I'm not sure this would work. Wouldn't all cases get overwritten by `MQTTPubcomp` after the fallthrough? I think I'll make it a separate helper function, something like `_getAckFromPacketType`, since the second switch increases complexity a bit.

Reply:
I was suggesting to add common code after  the switch actually. ( You can create  bool ackTypeSet ).  Helper function is probably better option though, will reduce complexity as well.